rule all: #this is to start with that file (in terminal: snakemake -n)
    input:
        "plots/quals.svg"


#Step 1: DESCOMPRESS tic
rule descompress:
    input:
        "data/compressed_samples/{samples}.fastq.gz"
    output:
        "data/samples/{samples}.fastq"
    shell:
        "gunzip -c {input} > {output}" 
#Step 2: QUALITY CONTROL tic
rule quality_control:
    input:
        "data/samples/{samples}.fastq"
    output:
        "data/samples/{samples}_fastqc.html"
    shell:
        "fastqc {input} | mv {output} > results/reports"

#Step 3: Mapping
rule map_reads:
    input:
        "data/GCA_904849725_genome.fa",
        "data/samples/{sample}.fastq"
    output:
        "results/mapped/{sample}.bam"
    conda:
        "envs/mapping.yaml"
    shell:
        "bwa mem {input} | samtools view -b - > {output}"
#Step 4: Sort .bam reads
rule sorted_bam:
    input:
        "results/mapped/{sample}.bam"
    output:
        "results/sorted_reads/{sample}.bam"
    shell:
        "samtools sort -T results/sorted_reads"
        "-O bam {input} > {output}"

